pkg/controller/apiextensions/composite/composed.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package composite
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+)
+
+// Observation is the result of composed reconciliation.
+type Observation struct {
+	Ref               v1.ObjectReference
+	ConnectionDetails managed.ConnectionDetails
+}
+
+// NewAPIComposedReconciler returns a new *APIComposedReconciler.
+func NewAPIComposedReconciler(c client.Client) *APIComposedReconciler {
+	return &APIComposedReconciler{
+		client: resource.ClientApplicator{
+			Client:     c,
+			Applicator: resource.NewAPIPatchingApplicator(c),
+		},
+	}
+}
+
+// APIComposedReconciler is able to reconcile a composed resource.
+type APIComposedReconciler struct {
+	client resource.ClientApplicator
+}
+
+// Reconcile tries to bring the composed resource into the desired state. It
+// creates the resource if the given reference is empty.
+func (r *APIComposedReconciler) Reconcile(ctx context.Context, cr resource.Composite, composedRef v1.ObjectReference, tmpl v1alpha1.ComposedTemplate) (Observation, error) {
```

Comment:
Its type is `Composition` right now but we plan to have different kinds of `Composition`s. At that point, it would be simply copy of `v1.ObjectReference`, which I'd not prefer to do.

Reply:
I actually think we should make copies of that type, for the reasons outlined in https://github.com/crossplane/crossplane-runtime/issues/49.